Add SpringController valve command

andermi opened the pull request:
Addresses #44 Valve command

Test with:
`ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 10\`

Can watch `/sc_record/range_finder` in rqt_plot. Value should rise ~0.25 m in 10 sec

Can close the valve with:
`ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 0\`

---

Modeling the valve (need to clean up w/ parameter):

```
void PolytropicPneumaticSpring::openValve(
  const int dt_nano,
  double & P1, double & V1,
  double & P2, double & V2)
{
  // want piston to drop 1 inch per second
  // so let mass flow from lower chamber to upper
  // results in initial pressure change (P0 -- or P1, P2 with hysteresis) via ideal gas law
  double dt_sec = dt_nano * 1e-9;
  static const double absement{5e-6};  // measure of valve opening cross-section and duration
                                       // units of meter-seconds
  double mass_flow = absement*this->dataPtr->P;  // P in Pascals (kg/(m*s^2)) so, mass_flow in kg/s
  double delta_mass = dt_sec*mass_flow;  // kg of gas per step
  if (this->dataPtr->is_upper)
  {
    this->dataPtr->mass += delta_mass;
    P1 = this->dataPtr->mass * this->dataPtr->R * this->dataPtr->T0 / V1;
    P2 = this->dataPtr->mass * this->dataPtr->R * this->dataPtr->T0 / V2;
  }
  else
  {
    this->dataPtr->mass -= delta_mass;
    P1 = this->dataPtr->mass * this->dataPtr->R * this->dataPtr->T0 / V1;
    P2 = this->dataPtr->mass * this->dataPtr->R * this->dataPtr->T0 / V2;
  }
}
```

opening for 5 seconds using component:
```
spring_state_comp->Data().valve_command = true;
spring_state_comp->Data().command_duration = 5s;
```

![Screenshot from 2022-05-27 11-37-07](https://user-images.githubusercontent.com/20071132/170775039-f18fbcd7-44c1-4d42-8c40-7b8ff499c36a.png)

lowers piston from 0.72 to 0.85 meters (`range_finder` value measured from top of upper chamber) in ~5.5 seconds
results in dropping 0.0236 (~1 inch) per second

Todo:

- [x] Model valve
- [x] Add command service to SpringController

hamilton8415 commented:
Looks good, nice way to characterize mass flow as a function of pressure differential  abasement.  absement is a new one for me:  http://wearcam.org/absement/examples.htm  This same technique will apply to the pump command as well, different sign and magnitude of course...

It looks like you have test code in there to control the time this valve is active, I'll be watching to see how you integrate the command and timeouts from the ROS2 spring controller code and will copy for similar behaviors in ElectroHydraulicPTO.  At the moment I simply respond to ignition transport callouts and maintain a time-out in the plugin.

andermi commented:
> integrate the command and timeouts

I made use of the [ignition::math::Stopwatch](https://gazebosim.org/api/math/6.10/Stopwatch_8hh_source.html) in my SpringState component in the ECM

```
  ignition::math::Stopwatch command_watch;
  ignition::math::clock::duration command_duration;
  bool valve_command{false};
  bool pump_command{false};
```

and manage the stopwatch like this:

```
manageCommandTimer(SpringState & state)
{
  if (state.valve_command || state.pump_command) {
    if (!state.command_watch.Running()) {
      ignerr << "Valve opening (" << \
        std::chrono::duration_cast<std::chrono::seconds>(state.command_duration).count() << \
        "s)" << std::endl;
      state.command_watch.Start(true);
    } else {
      if (state.command_watch.ElapsedRunTime() >= \
        state.command_duration)
      {
        state.command_watch.Stop();

        if (state.valve_command) {
          state.valve_command = false;
        }

        if (state.pump_command) {
          state.pump_command = false;
        }
        ignerr << "Valve closed" << std::endl;
      }
    }
  } else if (!state.valve_command && !state.pump_command) {
    if (state.command_watch.Running()) {
      state.command_watch.Stop();
      ignerr << "Valve closed" << std::endl;
    }
  }
}
```

andermi commented:
I added a `BUSY` flag in the command service response to indicate a command is already running to not issue a new valve command when we are already running the valve or pump. @hamilton8415 should the behavior instead be to preempt the current command? (e.g. stop the pump if I issue a valve, or extend the valve time if I issue another valve)

hamilton8415 commented:
In both the pump and valve commands, the behavior of the hardware is a new command sets the timeout to the new specified value.  So if you send a "valve 10" and then 5 seconds later send another "valve 10", the valve will be open for a total of 15 seconds.  Same for the pump.

andermi commented:
Ok, I'll modify the behavior then.

chapulina commented on buoy_gazebo/src/PolytropicPneumaticSpring/PolytropicPneumaticSpring.cpp:
```diff
@@ -123,6 +127,81 @@ double SdfParamDouble(
+        }
+      }
+    }
+  } else if (!state.valve_command && !state.pump_command) {
```
I think this may be the same as

```suggestion
  } else {
```

Just checking if that was the intention. I can see an argument for being explicit, but for someone trying to follow the logic it may add some complexity.

andermi replied:
I was being explicit (zen of python? :stuck_out_tongue_winking_eye: ) but I actually just made a pass at cleaning up logic and this went away.

chapulina commented on buoy_gazebo/src/controllers/SpringController/SpringController.cpp:
```diff
@@ -139,36 +146,55 @@ struct SpringControllerPrivate
   }
-
-/*  TODO(andermi) set up other services like this
+///*
```
This line could be removed instead of commented out

andermi replied:
oops! debugging leftovers :stuck_out_tongue_winking_eye:

chapulina commented on buoy_gazebo/src/controllers/SpringController/SpringController.cpp:
```diff
@@ -67,6 +70,10 @@ struct SpringControllerPrivate
 
+  std::atomic<bool> valve_command_{false}, pump_command_{false};
+  std::atomic<bool> has_new_command_{false}, reset_watch_{false};
+  ignition::math::clock::duration command_duration_;
```
I was surprised to see that `ignition::math::clock` was a thing 🧐 No action needed here, I'd just like to mention that I've ticketed this upstream:

* https://github.com/gazebosim/gz-math/issues/436

andermi replied:
glad I could help :laughing:

chapulina reviewed (commented):
I did a high-level review and didn't look into the math or output behaviour.

> Can close the valve with:
ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 0\

Should I see a message like "Valve closed" printed when I run this command?

andermi commented on buoy_gazebo/src/PolytropicPneumaticSpring/PolytropicPneumaticSpring.cpp:
```diff
@@ -153,6 +234,11 @@ void PolytropicPneumaticSpring::Configure(
   ignwarn << "name: " << this->dataPtr->name << std::endl;
 
+  this->dataPtr->is_upper = _sdf->Get<bool>("is_upper");
+  ignwarn << "is upper? " << std::boolalpha << this->dataPtr->is_upper << std::noboolalpha <<
```
oops, was just using these as debug and never changed them to actual debug

andermi commented on buoy_gazebo/src/PolytropicPneumaticSpring/PolytropicPneumaticSpring.cpp:
```diff
@@ -254,15 +340,22 @@ void PolytropicPneumaticSpring::Configure(
   const ignition::gazebo::UpdateInfo & _info,
   ignition::gazebo::EntityComponentManager & _ecm)
 {
-  IGN_PROFILE("PolytropicPneumaticSpring::PreUpdate");
```
I removed it since I didn't think we were using it. I'll put it back :)

chapulina replied:
Here's the profiler in action: https://github.com/osrf/buoy_sim/pull/52

andermi commented:
> Should I see a message like "Valve closed" printed when I run this command?

I changed these to ROS2 console prints now.

In the latest commits, I modified behavior to follow what @hamilton8415 mentioned. Also, cleaned up the command logic, and addressed PR comments.

Behavior when issuing console commands (don't mind the truncated duration errors for printing):
```
[ign gazebo-1] [INFO] [1654030769.374244685] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030769.374641468] [spring_controller]: Valve open (10s)
[ign gazebo-1] [INFO] [1654030773.070602607] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (0s)
[ign gazebo-1] [INFO] [1654030773.071846096] [spring_controller]: Valve closed after (3s)
[ign gazebo-1] [INFO] [1654030787.287722168] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030787.288250843] [spring_controller]: Valve open (10s)
[ign gazebo-1] [INFO] [1654030790.001460840] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030790.002160845] [spring_controller]: Valve open (12s)
[ign gazebo-1] [INFO] [1654030792.750108155] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030792.750821277] [spring_controller]: Valve open (15s)
[ign gazebo-1] [INFO] [1654030802.751982084] [spring_controller]: Valve closed after (15s)
```

andermi commented:
> Should I see a message like "Valve closed" printed when I run this command?

I changed these to ROS2 console prints now.

In the latest commits, I modified behavior to follow what @hamilton8415 mentioned. Also, cleaned up the command logic, and addressed PR comments.

Behavior when issuing console commands (don't mind the truncated duration errors for printing):
```
[ign gazebo-1] [INFO] [1654030769.374244685] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030769.374641468] [spring_controller]: Valve open (10s)
[ign gazebo-1] [INFO] [1654030773.070602607] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (0s)
[ign gazebo-1] [INFO] [1654030773.071846096] [spring_controller]: Valve closed after (3s)
[ign gazebo-1] [INFO] [1654030787.287722168] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030787.288250843] [spring_controller]: Valve open (10s)
[ign gazebo-1] [INFO] [1654030790.001460840] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030790.002160845] [spring_controller]: Valve open (12s)
[ign gazebo-1] [INFO] [1654030792.750108155] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[ign gazebo-1] [INFO] [1654030792.750821277] [spring_controller]: Valve open (15s)
[ign gazebo-1] [INFO] [1654030802.751982084] [spring_controller]: Valve closed after (15s)
```

chapulina commented:
> if you send a "valve 10" and then 5 seconds later send another "valve 10", the valve will be open for a total of 15 seconds

I think real time and sim time are getting mixed somewhere.

I'm trying to run this scenario like this:

1. Start simulation paused:
    `ign gazebo mbari_wec.sdf -v 3`
3. Send a command to open for 10s:
    `ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 10`
4. Step the simulation for 5 s (5000 iterations)
    `ign service -s /world/world_demo/control --reqtype ignition.msgs.WorldControl --reptype ignition.msgs.Boolean --timeout 1000 --req "pause: true, step: true, multi_step: 5000"`
5. Send a command to open for 10s:
    `ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 10`
    It printed `[INFO] [1654044735.192398374] [spring_controller]: Valve open (54s)`, I think that's counting real time, not sim time
6. Step for 5 seconds more:
    `ign service -s /world/world_demo/control --reqtype ignition.msgs.WorldControl --reptype ignition.msgs.Boolean --timeout 1000 --req "pause: true, step: true, multi_step: 5000"`
    It printed
    ```
    [Msg] Valve closed after (103s)
    [INFO] [1654044794.627076135] [spring_controller]: Valve closed after (103s)
    ```
    But we're at 10 s of simulation, I'd expect the valve not to close until 15 s.

andermi commented:
> I think real time and sim time are getting mixed somewhere.

Is this an issue with using [StopWatch](https://gazebosim.org/api/math/6.10/Stopwatch_8hh_source.html)?

andermi commented:
> Is there a timer that follows sim time?

@chapulina I made my own :nerd_face: 

your test should now work with sim time:
`$ ros2 launch buoy_gazebo mbari_wec.launch.py`
(need launch file for sim clock bridge from gz to ros2)

```
$ ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 10
requester: making request: buoy_msgs.srv.ValveCommand_Request(duration_sec=10)

response:
buoy_msgs.srv.ValveCommand_Response(result=buoy_msgs.msg.PBCommandResponse(value=0))
```
```
[ROS 2 Spring Control] ValveCommand Received (10s)
Valve open (10s)
```

step sim ~5sec:
```
$ ign service -s /world/world_demo/control --reqtype ignition.msgs.WorldControl --reptype ignition.msgs.Boolean --timeout 1000 --req "pause: true, step: true, multi_step: 5000"
data: true
```

send another command:

```
$ ros2 service call /valve_command buoy_msgs/srv/ValveCommand duration_sec:\ 10
requester: making request: buoy_msgs.srv.ValveCommand_Request(duration_sec=10)

response:
buoy_msgs.srv.ValveCommand_Response(result=buoy_msgs.msg.PBCommandResponse(value=0))
```
```
[ROS 2 Spring Control] ValveCommand Received (10s)
Valve open (15s)
```

step again ~10sec:

```
$ ign service -s /world/world_demo/control --reqtype ignition.msgs.WorldControl --reptype ignition.msgs.Boolean --timeout 1000 --req "pause: true, step: true, multi_step: 10000"
data: true
```
`Valve closed after (15s)`

chapulina commented on buoy_gazebo/src/PolytropicPneumaticSpring/SpringState.hpp:
```diff
@@ -31,6 +78,10 @@ struct SpringState
 
   // Commands
+  // ignition::math::Stopwatch command_watch;
+  // ignition::math::clock::duration command_duration;
```
These comments could be removed

andermi replied:
removed

chapulina commented on buoy_gazebo/src/buoy_utils/StopwatchSimTime.hpp:
```diff
@@ -0,0 +1,182 @@
+  /// has not been started.
+
+public:
+  rclcpp::Time StartTime() const;
```
Can this be returned as a const ref not to force a copy?

```suggestion
  const rclcpp::Time & StartTime() const;
```

The same for other getters below

andermi replied:
Yeah, I think so -- good optimization. I didn't look too closely at the Stopwatch API when I copied it from ignition-math.

chapulina replied:
Ahh yeah it looks like the elapsed times may return locally scoped variables, so we'll need to live with a copy there

chapulina commented on buoy_gazebo/src/buoy_utils/StopwatchSimTime.hpp:
```diff
@@ -0,0 +1,182 @@
+  StopwatchSimTime & operator=(StopwatchSimTime && _watch);
+
+#ifdef _WIN32
+// Disable warning C4251 which is triggered by
```
I'll mention something just in case for the future. We've been migrating our PIMPL pointers to [ignition::utils::ImplPtr](https://github.com/gazebosim/gz-utils/blob/ign-utils1/include/ignition/utils/ImplPtr.hh) so we don't need to workaround `std::unique_ptr`'s limitations.

andermi replied:
Another case of just copy/pasting from ignition-math's [Stopwatch.hh](https://gazebosim.org/api/math/6.10/Stopwatch_8hh_source.html). Should I remove these directives?

chapulina replied:
Yeah we haven't migrated `gz-math` to the new pattern yet. We've been trying to use `ImplPtr` for all new code, but honestly we often forget to 🙃 

They don't hurt, they just prevent warnings on Windows. We're not really targeting Windows with this project, so I'm fine either way.

chapulina commented on buoy_gazebo/src/buoy_utils/StopwatchSimTime.hpp:
````diff
@@ -0,0 +1,182 @@
+///   timeSys.ElapsedRunTime()).count() << " ms\n";
+/// watch.Stop();
+/// ```
+class StopwatchSimTime
````
Nice! This looks like something that could be generally useful. We should consider moving it somewhere like `ros_ign_gazebo`.

andermi replied:
It should also work with pure ROS2 using `/clock` published from any source (and `use_sim_time` param set per node). May want to put it in `rclcpp` somewhere?

andermi replied:
Actually, should work in ROS2 with any time source really. It just takes an instance of an `rclcpp::clock` from the `rclcpp::Node`.

chapulina replied:
Yeah true this is way more generic than I originally thought. I'm not sure what the best place would be, the `rclcpp` repo already looks very large. I'll ask around

chapulina reviewed (approved):
Works for me! Nice ROS-based sim time stopwatch!
